### PR TITLE
1232: Allow filtering options by a previous question

### PR DESCRIPTION
**Pull request by chris-bes**
### Changes:

- Allow filtering options by previous question

### CODE REVIEW

File: packages/meditrak-app/app/database/schema.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ import * as DataTypes from './types';
 
 export const schema = {
   schema: Object.values(DataTypes),
-  schemaVersion: 17,
+  schemaVersion: 18,
```
Comment (chris-bes) [inline packages/meditrak-app/app/database/schema.js:10]: I also increased the version in https://github.com/beyondessential/tupaia/pull/1451. Should I maybe branch off from it?

Comment (edmofro) [reply packages/meditrak-app/app/database/schema.js:10]: I think it'll be fine - could you just keep an eye to make sure they end up in the same app release?

File: packages/meditrak-app/app/assessment/specificQuestions/AutocompleteQuestion.js [hunk 2]
```diff
@@ -42,15 +47,29 @@ export class AutocompleteQuestion extends React.Component {
 
   buildOptionList = options => options.map(option => option.label || option.value);
 
+  getOptionAttributeFilters = () => {
+    const { attributeAnswers } = this.props;
+
+    return option =>
+      attributeAnswers
+        ? Object.entries(attributeAnswers).every(([key, answer]) => {
+            // No answer was selected for the question to filter, return all
```
Comment (edmofro) [inline packages/meditrak-app/app/assessment/specificQuestions/AutocompleteQuestion.js:56]: Is this comment in the right place?

File: packages/meditrak-server/src/routes/importSurveys/ConfigImporter/processAutocompleteConfig.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
+/**
```
Comment (edmofro) [inline packages/meditrak-server/src/routes/importSurveys/ConfigImporter/processAutocompleteConfig.js:1]: In here there's some quite familiar code - is there stuff we can reuse in a common util from your earlier PRs?

**Review by edmofro — CHANGES_REQUESTED** [review]
Looking good @chris-bes :-)

How's the performance of this? Any lag as you type a search filter on a big option set? It has to run the attributes filter across every option as well, including running the `toJson` every time. Shouldn't be slow, but with a large number of options...maybe?

**Comment by chris-bes** [comment]
> Looking good @chris-bes :-)
> 
> How's the performance of this? Any lag as you type a search filter on a big option set? It has to run the attributes filter across every option as well, including running the `toJson` every time. Shouldn't be slow, but with a large number of options...maybe?

So far no but I only tested with a small number of options. Will test it with more options and let you know!

File: packages/meditrak-server/src/routes/utilities/translateQuestionDependentFields.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import { splitStringOn } from './split';
+
+// JSON subfields will be in the format field.subfield, e.g. 'attributes.type'
+const isJsonSubfield = (jsonFieldList, fieldKey) =>
+  jsonFieldList.find(field => fieldKey.startsWith(field));
+
+export const translateQuestionDependentNonJsonFields = (config, nonJsonFieldTranslation) => {
```
Comment (chris-bes) [inline packages/meditrak-server/src/routes/utilities/translateQuestionDependentFields.js:7]: I moved the code that translate/replace the config into 1 util file like this to reduce code duplication. However, it felt a bit awkward for me to translate both non json (eg: parent, grandparent) and json (attributes.type, attributes.parent_project) in 1 method. So I split it into 2 different methods, 1 for translating non json and 1 for translating json which I feel is a bit cleaner (although it have a little bit of code duplication)

**Comment by chris-bes** [comment]
> Looking good @chris-bes :-)
> 
> How's the performance of this? Any lag as you type a search filter on a big option set? It has to run the attributes filter across every option as well, including running the `toJson` every time. Shouldn't be slow, but with a large number of options...maybe?

I've tested it with 100 options and searching was fast and no lag at all!

File: packages/meditrak-server/src/routes/utilities/translateQuestionDependentFields.js (hunk 4 above)
Comment (edmofro) [reply packages/meditrak-server/src/routes/utilities/translateQuestionDependentFields.js:7]: Good call. I would remove the `NonJson` bit from the function and variable names though. Having the other one include `Json` is enough to distinguish the two.

Comment (edmofro) [reply packages/meditrak-server/src/routes/utilities/translateQuestionDependentFields.js:7]: Actually I would use the term `Nested` rather than `Json`...

**Review by edmofro — APPROVED** [review]
Approved as is, with some optional name tweak suggestions :-)

**Pull request merged**